pep-0635.rst

```diff
@@ -0,0 +1,689 @@
+pattern matching targets Python objects rather than textual data,
+and builds on *decision trees* as the driving motor for finding a
+match.  Moreover, pattern matching blends the matching of a suitable
+pattern with that of a function, i.e. code that is executed in order
+to handle and process the information provided by a specific kind of
+object.
+
+One of the simplest forms of pattern matching that we find in other
+languages comes in the form of *function overloading*.  The type and
+number of arguments then determine which implementation of a specific
+function will be executed.  Object-oriented languages (including
+Python) may also use the *visitor pattern* to differentiate an action
+based on the type or class of an object.  Both of these approaches,
+however, are aimed at 'shallow' structures with little or no direct
+support for nested structures or structural information that is not
+directly encoded in an object's class or type.  For instance, it is
+simple to differentiate between an integer, a string, and a tuple, say,
+but becomes quite cumbersome and difficult to differentiate between
+tuples of different lengths, or between one containing string elements
+vs. one containing numeric elements.  This is where pattern matching
+shines: for structures that go beyond simple class distinctions.
+
+Although pattern matching is a concept that has been known and used
+for decades, we propose a re-interpretation that centres around the
+principle of 'duck typing' and builds on existing features of the
+Python language such as iterable unpacking.  Patterns adopt the syntax
+of parameters as far as possible and, to a somewhat lesser degree,
+that of targets in iterable unpacking.  In contrast to iterable
+unpacking, pattern matching is a 'conditional' feature that has to
+avoid side-effects, i.e. extracting elements from an abstract iterable
+(thus working with actual sequences instead) or assigning to non-local
+targets such as object attributes or container elements.  Overall, we
+followed the guiding principle that patterns be static templates for
+the structure and type of objects, i.e. patterns should depend as
+little as possible on the surrounding context or current values of
+variables (other than the subject to be matched, that is).
+
+Pattern matching is a structure that *maps* different patterns/templates
+to 'function bodies' or actions.  This general mapping structure can be
+found in different context as well.  Algol-derived languages usually
+provide a switch table that maps ordinal values to actions, whereas Lisp
+has a more general mapping from general conditions to actions.  Although
+all these constructs share a similar overall structure and some syntax,
+their intents and motivation differs highly.  In particular, pattern
+matching as proposed here is not intended as or an extension of a switch
+structure, although it is possible to emulate it to a large degree with
```

Comment:
structure -> statement, emulate it -> emulate one.

Reply:
Agreed.